### PR TITLE
Enhance dovecot configuration layout

### PR DESCRIPTION
**Pull request by ganto**
So far the dovecot configuration layout is not very flexible. There are still a lot of "simple" setups which cannot be done with the current state of the dovecot role. I tried to find a way to improve this situation (see #8) and ended up with this suggestion.

The more flexible plugin support as implemented in my _ansible-dovecot:mail-plugins_ branch is already included here. However, the code so far is still missing the LMTP support.

There is one "nasty" hack in the jinja2 templates for which I'm still looking for an improvement. I'm trying to lookup optional(!) sub-keys in a dictionary (see [dovecot.conf.j2](https://github.com/debops/ansible-dovecot/blob/a27c0102899d2c6b2e7b3753b4c204a03e8b05ea/templates/etc/ferm/filter-input.d/dovecot.conf.j2#L31))

If there would be a way to concatenate variables to variable names in jinja, the templates could be made even shorter. However, I didn't figure out a way how to do that.

I'm curious for your comments.

### CODE REVIEW

**Comment by drybjed** [comment]
Looks much better, nice work. I haven't found where defaults are defined, but I haven't looked too closely, do you just rely on Dovecot internal defaults?

It looks like task list with diversions lacks `when` statements, all tasks will be executed over and over...

You should move default variable documentation to separate `docs/defaults-configuration.rst` config file. Look for examples at `debops.dhcpd` or `debops.mariadb` roles.

**Comment by ganto** [comment]
In comparison with the _ansible-dovecot:service-map_ branch, there are no explicit default configurations anymore. This simplifies the template generation as I don't need to merge nested dicts anymore. Except for the `allow` variable(s) where default is implicitly [set](https://github.com/debops/ansible-dovecot/pull/9/files#diff-2de66ee071faa076f8e8c5768b8b8f60R13) in the template, I rely on the dovecot internal defaults.

What do you mean with the condition for the config diverts? As I compile the file list only during the task run, I hardly can skip this task.

I don't really understand the build failure. Does this have to do with the comments explaining the dict layout? I'll definitely move that to the `docs/defaults-configuration.rst` (didn't notice this before).

There is still one planned feature that I'm struggling with. As you can see there are some [service lists](https://github.com/debops/ansible-dovecot/pull/9/files#diff-6d62b73b3f46cfa11f6466d08cf45303R3) in the `vars/main.yml`. With help of these lists, I resolve the listener configurations I have to query. I was imagining that someone could e.g. set something like this, to specify additional listeners (following the example in `/usr/share/dovecot/conf.d/20-managesieve.conf`):

    dovecot_managesieve_services: [ 'managesieve', 'sieve_deprecated' ]

    dovecot_managesieve_config_map:
        login-service:
            inet-listener:
                sieve_deprecated:
                    port: 2000 

This would still configure the default Managesieve listener and an alternative one on port 2000.

However, I only noticed now, that the variables from `vars` cannot easily be overwritten in DebOps using the inventory. Is this correct? Alternatively, I could move these variables to `defaults`. At the other side, this would add another bunch of variables to the `defaults` and I also worry, that users might mess with it without understanding the implication. Maybe a clearer name e.g. `dovecot_<protocol>_listeners` would help with that.

**Comment by drybjed** [comment]
OK, I've looked closer at the divect tasks and see that you add/remove config files based on what dovecot capabilities are enabled. Clever! :-)

Yes, build failure is due to the docs. `defaults/main.yml` is converted to `.rst` and resulting file has errors which are caught by the sphinx test on Travis. No more broken http://docs.debops.org/ :-) When you move that to `docs/` it should all be fine.

You cannot override variables in `vars/main.yml` of a particular role, not from inventory, not from other roles. So yeah, moving that to `defaults/main.yml` will be fine. If you split the default variables in sane sections I think that we will be OK with many variables there.

**Comment by ganto** [comment]
Ok, it's shaping up more and more. The doc error is fixed. The `vars` are gone again.

Now I still have to cleanup the `dovecot_protocols` vs. `dovecot_<protocol>_listeners` mess. I thought of enabling the protocol types (IMAP, POP3,...) still in `dovecot_protocols` but then the listeners (IMAP, IMAPS) in the `dovecot_<protocol>_listeners` list. This would mean:

Current (v0.1.0):

    dovecot_protocols = [ 'imap', 'imaps' ]

Proposal:

    dovecot_protocols = [ 'imap' ]
    dovecot_imap_listeners = [ 'imap', 'imaps' ]

On the other side I could also completely get rid of `dovecot_protocols` and only enable a protocol if `dovecot_<protocol>_listeners` is defined and not empty.

Personally I prefer to keep `dovecot_protocols` around. So ppl would ideally only need to set this if they need to turn on/off a protocol. Any suggestions?

**Comment by drybjed** [comment]
So, it looks like Ubuntu team will try and backport the fix to Trusty. In the meantime, you could try switching the mariadb_server_flavor to mysql, it should work fine then. After you test it, it can be added to the flavor map to fix the issue on Trusty for the time being.

**Comment by drybjed** [comment]
Sure, having a list of main protocols to enable/disable and then separate entry for details of each which can be modified as needed seems fine to me.

File: docs/defaults-configuration.rst [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+   :depth: 1
+
+.. _dovecot_protocols:
+.. _dovecot_imap_config_map:
+.. _dovecot_imap_listeners:
+.. _dovecot_pop3_config_map:
+.. _dovecot_pop3_listeners:
+.. _dovecot_lda_config_map:
+.. _dovecot_managesieve_config_map:
+.. _dovecot_managesieve_listeners:
```
Comment (drybjed) [inline docs/defaults-configuration.rst]: You probably want to move these to just before each section, then when somebody clicks on the reference, browser will jump to relevant section automatically.

**Comment by ganto** [comment]
I think i'm fine to merge it now. Will probably still find one or another minor issue, but should work quite well with all the different configs that I tried. Changelog and docs should be up to date.

If you (@drybjed) don't have anything to add anymore, feel free to push the button :-)

**Comment by drybjed** [comment]
@ganto Glad to hear it. Thanks for your awesome work on this role! I hope that will now be a good base to extend the functionality as needed. I'll tag that version as well so it's easier to track.

**Comment by ganto** [comment]
Thanks for the merge... will look into LMTP support now :-)

**Pull request merged**